### PR TITLE
[t2386] WIP - hardening dependencies.js to not stumble when assets are m...

### PR DESCRIPTION
**Pull request by secretrobotron**
...issing

### CODE REVIEW

File: src/dependencies.js [hunk 1]
```diff
@@ -2,214 +2,53 @@
...
-                callback();
-              }
-            }
-          }, CSS_POLL_INTERVAL );
-        }
+    this.prepareLESSEnvironment = function( readyCallback, errorCallback ) {
+      // Load less.js so we can parse the *.less -> *.css
+      _loaders.js( _configDirs.tools + LESS_JS, true, readyCallback, function() {
+          return !!window.less;
+        }, errorCallback );
```
Comment (mjschranz) [inline src/dependencies.js]: I think these 2 lines should be bumped to the left 2 spaces

File: src/loaders/less-loader.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [ "util/xhr", "./base-loader", ], function( XHR, BaseLoader ) {
```
Comment (mjschranz) [inline src/loaders/less-loader.js]: lint doesn't catch this comma? Odd.

File: src/ui/ui.js [hunk 3]
```diff
@@ -121,11 +110,28 @@ define( [ "core/eventmanager", "./toggler",
...
+          loadOptions.type = "css";
+          loadOptions.url = BUTTER_CSS_FILE;
+          loadUI();
+        }
 
         _this.tray.attachToDOM();
         _this.header.attachToDOM();
       }
-      else{
+      else {
```
Comment (mjschranz) [inline src/ui/ui.js:120]: Could put this else on the line above for consistency with other changes in the file.

File: src/loaders/less-loader.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      if ( this.readyState === 4 ) {
+        var parser = new less.Parser({
+          paths: [
+            self.configDirs.css
+          ],
+          error: error
+        });
+        lessFile = this.response;
+
+        parser.parse( lessFile, function( e, root ){
```
Comment (mjschranz) [inline src/loaders/less-loader.js]: Might as well keep spacing consistent since these are new files.

File: src/loaders/less-loader.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+          if( e ){
```
Comment (mjschranz) [inline src/loaders/less-loader.js]: Same as above.

File: src/loaders/js-loader.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  JSLoader.prototype = Object.create( BaseLoader );
+  JSLoader.prototype.constructor = JSLoader;
+
+  JSLoader.prototype.load = function( url, exclude, callback, checkFn, error ) {
+    checkFn = checkFn || BaseLoader.DEFAULT_CHECK_FUNCTION();
+
+    url = this.fixUrl( url );
+
+    if( !checkFn() ){
```
Comment (mjschranz) [inline src/loaders/js-loader.js]: Spacing.

File: src/loaders/js-loader.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        // Opera has a bug that will cause it to also fire load after
+        // setting it to null to prevent this
+        scriptElement.onload = null;
+
+        error( e );
+      };
+      scriptElement.src = url;
+      document.head.appendChild( scriptElement );
+    }
+    else if( callback ){
```
Comment (mjschranz) [inline src/loaders/js-loader.js]: Spacing.

File: src/loaders/css-loader.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          if( callback ){
+            callback();
+          }
+        }
+      }, CSS_POLL_INTERVAL );
+    }
+
+    url = this.fixUrl( url );
+
+    if( !checkFn() ){
```
Comment (mjschranz) [inline src/loaders/css-loader.js]: Spacing.

File: src/loaders/css-loader.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      link = document.createElement( "link" );
+      link.type = "text/css";
+      link.rel = "stylesheet";
+      link.onerror = error;
+      link.onload = runCheckFn;
+      link.href = url;
+      document.head.appendChild( link );
+    }
+    else if( callback ){
```
Comment (mjschranz) [inline src/loaders/css-loader.js]: Spacing.

File: src/loaders/js-loader.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [ "./base-loader", ], function( BaseLoader ) {
```
Comment (mjschranz) [inline src/loaders/js-loader.js]: How does lint not catch this?!?!?!

File: src/plugin/module.js [hunk 11]
```diff
@@ -99,6 +99,7 @@ define( [ "core/logger", "./plugin-list", "./plugin" ],
           butter.dispatch( "pluginadded", plugin );
         }
         onReadyCallback();
+      }, function() {
```
Comment (mjschranz) [inline src/plugin/module.js:102]: Don't the loaders have other defaults in place? This looks ugly as is IMO.

File: src/dependencies.js [hunk 12]
```diff
@@ -2,214 +2,53 @@
...
-        if( !checkFn() ){
-          link = document.createElement( "link" );
-          link.type = "text/css";
-          link.rel = "stylesheet";
-          link.onerror = error;
-          link.onload = runCheckFn;
-          link.href = url;
-          document.head.appendChild( link );
+      errorCallback = errorCallback || function(){
+        if( console && console.log ){
```
Comment (mjschranz) [inline src/dependencies.js]: Spacing.

File: src/dependencies.js (hunk 1 above)
Comment (secretrobotron) [inline src/dependencies.js]: Actually, I put them like that on purpose. Thought it made it easier to read which params belongs to which calls.

Comment (secretrobotron) [inline src/dependencies.js]: Nvm. You're right. It's too tiny to matter.

File: src/plugin/module.js (hunk 11 above)
Comment (secretrobotron) [inline src/plugin/module.js:102]: Better? (see below)

File: src/ui/ui.js (hunk 3 above)
Comment (secretrobotron) [inline src/ui/ui.js:120]: Most of these are exploded in this file, actually. So, I fixed the other one I added instead. Good?

File: src/dependencies.js [hunk 13]
```diff
@@ -2,214 +2,53 @@
...
-        if( !checkFn() ){
-          link = document.createElement( "link" );
-          link.type = "text/css";
-          link.rel = "stylesheet";
-          link.onerror = error;
-          link.onload = runCheckFn;
-          link.href = url;
-          document.head.appendChild( link );
+      errorCallback = errorCallback || function() {
+        if ( console && console.log ) {
```
Comment (humphd) [inline src/dependencies.js]: We use console without checks elsewhere, why so cautious here?  If you really care, move this into shims.js so we can count on console and console.log/warn/error being callable.

File: src/loaders/base-loader.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [], function() {
+
+  var VAR_REGEX = /\{([\w\-\._]+)\}/;
```
Comment (humphd) [inline src/loaders/base-loader.js]: Needs description of what this is looking for.

File: src/loaders/js-loader.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+  function JSLoader( configDirs ) {
+    BaseLoader.call( this, configDirs );
+  }
+
+  JSLoader.prototype = Object.create( BaseLoader );
+  JSLoader.prototype.constructor = JSLoader;
+
+  JSLoader.prototype.load = function( url, exclude, callback, checkFn, error ) {
+    checkFn = checkFn || BaseLoader.DEFAULT_CHECK_FUNCTION();
```
Comment (humphd) [inline src/loaders/js-loader.js]: Do the same for error, since you call it below in some cases (i.e., null function so it doesn't throw if not provided).

Comment (humphd) [inline src/loaders/js-loader.js]: Same for callback, I guess.

File: src/loaders/load-group.js [hunk 16]
```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [ "./load-item" ], function( LoadItem ) {
+
+  function LoadGroup( loaders, readyCallback, errorCallback, ordered ) {
```
Comment (humphd) [inline src/loaders/load-group.js]: Add some simple docs to explain what a LoadGroup is.

File: src/loaders/load-group.js [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        var item = _items[ itemIndex++ ];
+        if ( item ) {
+          item.load( unorderedReadyCallback, unorderedErrorCallback );
+        }
+        else {
+          readyCallback( _this );
+        }
+      };
+
+      var unorderedReadyCallback = function( loadItem ) {
```
Comment (humphd) [inline src/loaders/load-group.js]: These can just be functions, you don't need to explicitly use var

```
function unorderedReadyCallback( loadItem ) {
...
```

File: src/loaders/load-group.js [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        _successfulItems.push( loadItem );
+        next();
+      };
+
+      var unorderedErrorCallback = function( loadItem ) {
```
Comment (humphd) [inline src/loaders/load-group.js]: Same here.

File: src/loaders/load-group.js [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    function startUnordered() {
+      var checkFinished = function() {
+        if ( _successfulItems.length + _erroneousItems.length === _items.length && readyCallback ) {
+          readyCallback( _this );
+        }
+      };
+
+      var unorderedReadyCallback = function( loadItem ) {
```
Comment (humphd) [inline src/loaders/load-group.js]: Same deal

File: src/loaders/load-group.js [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+        _successfulItems.push( loadItem );
+        checkFinished();
+      };
+
+      var unorderedErrorCallback = function( loadItem ) {
```
Comment (humphd) [inline src/loaders/load-group.js]: And here

File: src/loaders/load-item.js [hunk 21]
```diff
@@ -0,0 +1,38 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [], function() {
+  function LoadItem( type, loader, url, exclude, checkFunction ) {
```
Comment (humphd) [inline src/loaders/load-item.js]: Please document args and what you expect/mean by each.  Some are obvious, some (exclude) are not.

File: src/loaders/css-loader.js [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+  CSSLoader.prototype.constructor = CSSLoader;
+
+  CSSLoader.prototype.load = function( url, exclude, callback, checkFn, error ) {
+    var link,
+        interval,
+        img,
+        alreadyFired = false;
+
+    // Run the load function if the link variable hasn't already been initialized.
+    // TODO: Come up with a better check.
```
Comment (cadecairos) [inline src/loaders/css-loader.js:23]: We should file a bug on this so it doesn't get forgotten and then mention the number on this line

**Pull request merged**